File: src/videoCourse_03/lessons/lesson18/MyArrayList.java
package videoCourse_03.lessons.lesson18;

public class MyArrayList {

    private String[] array = new String[10];
    private int size = 0;

    public void add(String s) {
        array[size] = s;
        size++;
        if (size == array.length) {
            String[] newArray = new String[array.length * 2];
            for (int i = 0; i < array.length; i++) {
                newArray[i] = array[i];
            }
            array = newArray;
        }
    }

    public void remove(int index) {
        if (index >= 0 && index < size) {
            for (int i = index; i < size - 1; i++) {
                array[i] = array[i + 1];
            }
            size--;
        }
    }

    public void remove(String s) {
        int index = -1;
        for (int i = 0; i < size; i++) {
            if (s.equals(array[i])) {
                index = i;
            }
        }
        if (index != -1) {
            remove(index);
        }
    }

    public int getSize() {
        return size;
    }

    public String get(int index) {
        if (index >= 0 && index < size) {

            return array[index];
        } else {
            return "Элемент не найден";
        }
    }
}

class MyArrayListTest {
    public static void main(String[] args) {
        MyArrayList employees = new MyArrayList();
        employees.add("Иван");
        employees.add("Петр");
        employees.add("Семен");
        employees.add("Максим");
        employees.add("Денис");
        employees.add("Николай");
        for (int i = 0; i < employees.getSize(); i++) {
            System.out.println(employees.get(i));
        }
        System.out.println("------------------");
        employees.remove(2);
        for (int i = 0; i < employees.getSize(); i++) {
            System.out.println(employees.get(i));
        }
        System.out.println("------------------");
        employees.remove("Петр");
        employees.remove("Николай");
        for (int i = 0; i < employees.getSize(); i++) {
            System.out.println(employees.get(i));
        }
    }
}

File: src/videoCourse_04/kot/lessons/lesson415_generics/MyArrayList.kt
package videoCourse_04.kot.lessons.lesson415_generics

class MyArrayList : MyList<String> {

    private var capacity = 10
    private var array = Array(capacity) { "" }
    private var size = 0

    override fun get(index: Int): String {
        if (index in 0 until size) {
            return array[index]
        }
        throw ArrayIndexOutOfBoundsException()
    }

    override fun add(string: String) {
        if (size < capacity) {
            array[size] = string
        } else {
            capacity += 10
            val newArray = Array(capacity) { "" }
            for (i in array.indices) {
                newArray[i] = array[i]
            }
            array = newArray
            array[size] = string
        }
        size++
    }

    override fun remove(element: String) {
        for (i in array.indices) {
            if (array[i] == element) {
                size--
                if (i == size) {
                    array[i] = ""
                    break
                }
                for (j in i until size) {
                    if (j + 1 == size) {
                        array[j] = array[j + 1]
                        array[j + 1] = ""
                        break
                    } else {
                        array[j] = array[j + 1]
                    }
                }
            }
        }
    }

    override fun removeAt(index: Int) {
        if (index in 0 until size) {
            size--
            if (index == size) {
                array[index] = ""
            } else {
                for (i in index until size) {
                    if (i + 1 == size) {
                        array[i] = array[i + 1]
                        array[i + 1] = ""
                        break
                    } else {
                        array[i] = array[i + 1]
                    }
                }
            }
        } else {
            throw ArrayIndexOutOfBoundsException()
        }
    }

    override fun size(): Int = size
}